### ISSUE TITLE
Implement evolution ("morphing") for monsters

### ISSUE THREAD

**Issue by Sanglorian** [Tuxemon/Tuxemon#540]
Monsters should be able to evolve into new forms. 

**Levelling Up**
The main way for this to happen is through levelling up: a monster will have a level specified. When they gain a level and they are now equal to or greater than that specified level, they begin to evolve. The player has the opportunity to interrupt the evolution. This stops evolution taking place this time. 

**Temporary Badge Evolution**
Another way that's imagined by the Spyder in the Cathedral plot is a temporary evolution when an item is applied to the monster. This would only last until the party next rested/was healed. 

Badges are used up when used, and cost money.
Booster Tech is a badge replacement. It isn't used up when used. Because it substitutes for multiple badges, players will have to choose which evolution they want where multiple options exist. 

**Other**
A bunch of other ways of evolving are considered, including conditional based on gender, or special techniques known, or stat experience, or a bunch of other systems we haven't currently incorporated.

More discussion here: https://wiki.tuxemon.org/index.php?title=Morphing

Thanks team!

**Comment by Sanglorian** [comment]
I thought about this on the drive to work, and I think we could remove "Temporary Badge Evolution" and keep the same story benefit with a much simpler system: 

Some monsters evolve when a special item is used on them. Companies sell these items. But open source hackers have developed five open-source alternatives which are reusable, and therefore much cheaper. The evolution is permanent (the story could describe it as only lasting weeks or months to keep the "planned obsolescence" theme, but that'd just be fluff). 

This would remove the need to program de-evolution. Also instead of one open-source "Booster Tech" that would have to bring up all evolution options, you could just have five items, each of which triggers one of the options, if applicable. 

For example: 

Memnomnom evolves into Miaownolith or Pyraminx. Scoop Co sells "Miaownolith Milk" and "Sacred Keystone" for $1,000 each. Use one of those on your Memnomnom, and it evolves into one or the other.

Alternatively, if a hacker gives you the Metal Booster, you can use it on Memnomnom and it evolves into Pyraminx. Use the Earth Booster, and it evolves into Miaownolith. 

I think this is a lot more straightforward to program, and keeps the theme.

**Comment by MirceaKitsune** [comment]
In my book, evolutions are the next essential feature needed to get the engine closer to completion and allowing full functioning games to be built, now that exp and leveling up seem to be finally implemented. I don't know to what extent I can help with its implementation, this will require animations too which are not something I'm good at... still I might be able to lay down the foundations such as settings variables and functions, will have to look into this later.

**Comment by bitcraft** [comment]
It would be useful to get this started by drafting a file format which defines the monster evolutions.  There may be some existing JSON.  Once the draft is complete, more evolutions can be added and will give developers some data to work with.  Ill add a milestone to track progress.

**Comment by MirceaKitsune** [comment]
Would there need to be a separate JSON file to define evolutions? I figured we could just add two new parameters to every monster that uses morphing: A string specifying the name of the monster they're morphing into, and an integer defining at which level it happens.

**Comment by bitcraft** [comment]
thats a good point.  personally, i think a separate file makes more sense, especially if designers are working on just evolutions so that you can quickly define all the transformations without diving into a bunch of files, but i can see why it would also make sense to put evolutions into the monster files.  we should also consider the item transformations, and how we would store that in the files.

are we going to support different evolution paths?  something like eevee with the many ways to evolve?

**Comment by Sanglorian** [comment]
I think we should support different paths - they're a big part of many of the completed monsters. 

If possible, it would be good for the morphing data to just get exported out of the wiki along with the rest of the monster data. I think it would be possible to define it simply enough for that to be feasible.

Currently all the monsters that evolve/morph have just one of two triggers:

* They increase in level
* They have an item used on them

This may not be sufficient (e.g. they may also need to know a particular technique or be in a particular location). Also, there may be multiple triggers for a single morph. 

So I think morphing information could be limited to something that looks like this:

{ Memnomnom morph-option 
number: 1
destination: Miaownolith
trigger: Miaownolith Milk [an item that is consumed on use]
}

{ Memnomnom morph-option 
number: 2
destination: Miaownolith
trigger: Earth Booster [an item that is not consumed on use]
}

{ Memnomnom morph-option 
number: 3
destination: Pyraminx
trigger: Sacred Keystone [an item that is consumed on use]
}

{ Memnomnom morph-option 
number: 4
destination: Pyraminx
trigger: Metal Booster [an item that is not consumed on use]
}

Other examples: 

{ Vivipere morph-option
number: 1
destination: Vivisource
triggered by: level
level: 24
condition: male-gender 
}

(Obviously we haven't implemented gender yet and have no plans to - but that's true for a number of triggers).

{ Vivipere morph-option
number: 3
destination: Vivitron
triggered by: level
level: 25
}

**Comment by airon90** [comment]
> So I think morphing information could be limited to something that looks like this:

By now I would add "triggered by: item" in the first cases and instead of "level" I would say just "trigger" in order to uniform keys

**Comment by MirceaKitsune** [comment]
I've been thinking of giving a shot to implementing the basics for this. No animations or complex features for now: Just add a morphing.json file, teach the code to load it, then whenever the level_up function is called for a monster scan it and if your monster reached the level defined there just remove it and add the replacement monster in its place (with the same name and level).

Branching evolutions will likely have to be a thing of the future after such simple evolutions are added first; They would require a dialog to ask the player what choice they wish to pick after battle, don't think I can dive that far into it but this would provide the foundation.

I'm only concerned about one thing: The monster.level_up function is called during battle. If I remove the player's active monster while a battle is still going, there's definitely going to be breakage! Hopefully I can work around this by doing the check and replacement in the the function that runs when the battle ended... but even then the existing monster's sprites are still being drawn, so I should do this after the battle screen fades to black and the overworld loads again, it would help if anyone could point me to where in the code such a function is located with access to the player object.

**Comment by MirceaKitsune** [comment]
Looked a bit through this code just now. It feels like adding a whole new database category to address monsters would be rather convoluted, compared to just adding a new field in each monster's existing config. Would it be okay to go with this approach in the end? It seems both easier to implement as well as cleaner to define and maintain.

**Comment by Sanglorian** [comment]
Hey MirceaKitsune - thanks for working on that! 

> Branching evolutions will likely have to be a thing of the future after such simple evolutions are added first; They would require a dialog to ask the player what choice they wish to pick after battle, don't think I can dive that far into it but this would provide the foundation.

I don't think we'll need a dialogue box for branching evolutions, as long as there's a priority order. Upon levelling up, monsters can begin to evolve into the first evolution that they meet the criteria for.  Of course that can get implemented later too -- just saying I don't think there needs to be any dialogue option. 

> No animations or complex features for now: Just add a morphing.json file, teach the code to load it, then whenever the level_up function is called for a monster scan it and if your monster reached the level defined there just remove it and add the replacement monster in its place (with the same name and level).

Sounds good! One thing the animation does is gives the player time to cancel the evolution if they don't want it. If you're including text at this point, e.g. "NAME is morphing", that could be the point where pressing the cancel button would cancel the evolution.

---

I've put together some notes about the procedure and a potential animation, based on one put together by tamashihoshi. Not suggesting that these need to be implemented right now, of course - but I hope it helps people plan.

**Procedure **

_After battle_

- At the end of each battle the player wins, check each of the player’s monsters that increased in level during that battle. 
- If one of them has a morph-option specified that is “triggered by: level”, check the “trigger” for that morph option. 
- If the number in the trigger is less than or equal to the monster’s level, check any further conditions. 
- If those conditions are also met, play the animation.  
- If the player presses “cancel” at any point during the animation, end this process and check the next eligible monster, if any. The current monster does not morph. 
- Otherwise, delete the monster and replace it with the monster listed as “destination” under the morph-option. The new monster has the same position in the party, experience point total and custom name (if any) as the old monster. 
- Check the next eligible monster, if any.

_When an item is used_

- If an item from the Morph Items category is used on a monster, check if that monster has a morph-option specified that is "triggered by: item".
-- If it doesn't, show the dialogue: "It doesn't seem to respond."
- If it is triggered by item, check if the "trigger" for that morph option is the item's name.
-- If it doesn't, show the dialogue: "It doesn't seem to respond to that item."
- If the item does match the trigger, check any further conditions. 
- If those conditions are also met, play the animation.  
- If the player presses “cancel” at any point during the animation, end this process and check the next eligible monster, if any. The current monster does not morph. 
- Otherwise, delete the monster and replace it with the monster listed as “destination” under the morph-option. The new monster has the same position in the party, experience point total and custom name (if any) as the old monster. 

**Potential animation**

Mock up: https://wiki.tuxemon.org/index.php?title=File:Jl2ESeg.gif

- Fade in from black to:
- A background with the Chinese character for the Type that monster belongs to, along with the front sprite of that monster.
- The dialogue: “NAME is morphing!” appears. The monster’s Cry sounds.
- Flash of white.
- Music plays. An animation related to the Type plays, as the monster shakes, eventually hiding the monster. 
- Flash of white.
- The Type animation is replaced with a still GIF related to the Type, plus the front sprite of the new morphed form. The new monster’s Cry sounds. 
- A final sound effect plays. Dialogue shows based on type:
-- “NAME rose from the rubble” (Earth type)
-- “NAME emerged from the flames”
-- “NAME formed from the water”
-- “NAME was crafted” (Metal type)
-- “NAME has sprouted” (Wood type)
- Fade to black.

**Comment by MirceaKitsune** [comment]
Nice ideas. They'll definitely be for later though, not something I know how to implement personally. I plan on looking into it this only the basics: Define a set of morph options per monster, read them whenever that monster levels up, and if true then replace the monster. Once this foundation is there the rest can be slowly added in over time.

I was busy today but am also waiting on the devs to confirm that it's okay to implement this in the monster json instead. I looked at the database system and it seems more complicated to define a whole new category of databases, just to store options which address monsters anyway so they can be stored with all their other properties.

Also I'd like to know where the function that ends the battle state and reactivates the overworld is. I'd like to do the check right after the battle fades to black and before the overworld fades back in.

**Comment by MirceaKitsune** [comment]
Actually I have an additional idea in mind... would like to know what @bitcraft and other devs think on this too: What if we don't automatically evolve the monster after battle, but instead have an event action do it? That way if you don't want evolutions to be immediate but instead require the player to bring their monster to say a device that evolves them, you can have your own triggers for this... whereas if do, simply execute the evolution action right after the battle action (this might make event chains for random encounters a bit more tedious to maintain though).

The evolution itself would still be stored as a property by the code. So when monster.level_up is called, it sets a string variable on the monster which indicates what they need to morph into. However that morphing only actually happens when the event action for it is called.

**Comment by MirceaKitsune** [comment]
Merge request for evolutions is up! Apart from no animations which can be added later, it fully implements the functionality for tuxemon evolutions. Once it's merged into development this MR will be followed by another implementing the evolutions for the default game. See #585 for more info.
